google/cloud/storage/internal/tuple_filter.h

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_TUPLE_FILTER_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_TUPLE_FILTER_H_
+
+#include "google/cloud/internal/disjunction.h"
+#include "google/cloud/internal/invoke_result.h"
+#include "google/cloud/internal/tuple.h"
+#include "google/cloud/internal/utility.h"
+#include "google/cloud/storage/version.h"
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+
+/**
+ * A helper class to filter a single element from a tuple.
+ *
+ * Depending on whether to filter the element of not, appropriate tuple type is
+ * generate and a "filtering" member function, which either returns an empty
+ * tuple or a tuple containing the argument.
+ *
+ * @tparam T of the filtered element
+ * @tparam Filter whether to filter the element.
+ */
+template <typename T, bool Filter>
+struct TupleFilterItem {};
+
+/**
+ * Implementation of TupleFilterItem - true branch.
+ */
+template <typename T>
+struct TupleFilterItem<T, true> {
+  using Result = std::tuple<T>;
+  Result operator()(T&& t) const { return Result(std::forward<T>(t)); }
```

Comment:
FWIW, since `T` is class-level template parameter, it will be already bound when this function is called, and so the parameter type `T&&` here is not a forwarding reference, it's just a regular rvalue. And since this is just an rvalue, there's no need to use `std::forward`, we should just use `std::move` instead.

However, if instead we want this `op()` to use a forwarding reference, I think we'll need to make this member function itself a template and have it accept a parameter of type, say, `U&&`. That would be a forwarding reference so we could then use `std::forward`.

In general, I think we should only use `std::forward` with actual forwarding references, and use `std::move` otherwise. I think there are a other cases below that use `std::forward` with regular rvalues that we might want to look at again also.

Reply:
Ooops. Fixed now.